File: python/command/command.py
from abc import ABC


class ICommand(ABC):
    def execute(self):
        pass


class NoCommand(ICommand):
    def execute(self):
        pass


class RemoteController:
    def __init__(self):
        self._onCommand: ICommand = NoCommand()
        self._offCommand: ICommand = NoCommand()

    def set_on_command(self, command: ICommand):
        self._onCommand = command

    def set_off_command(self, command: ICommand):
        self._offCommand = command

    def press_on_button(self):
        self._onCommand.execute()

    def press_off_button(self):
        self._offCommand.execute()


class Bulb:
    def light_on(self):
        print("Light On!")

    def light_off(self):
        print("Light Off!")


class LightOnCommand(ICommand):
    def __init__(self, bulb: Bulb):
        self._bulb = bulb

    def execute(self):
        self._bulb.light_on()


class LightOffCommand(ICommand):
    def __init__(self, bulb: Bulb):
        self._bulb = bulb

    def execute(self):
        self._bulb.light_off()


def main():
    bulb = Bulb()
    controller = RemoteController()
    controller.set_on_command(LightOnCommand(bulb))
    controller.set_off_command(LightOffCommand(bulb))
    controller.press_on_button()
    controller.press_off_button()


if __name__ == '__main__':
    main()

File: cpp/command/command.cpp
#include <iostream>
#include <memory>
#include <utility>

class ICommand {
public:
    virtual void execute() = 0;
};

class RemoteController {
public:
    void set_on_command(std::shared_ptr<ICommand> command) {
        _on_command = std::move(command);
    }

    void set_off_command(std::shared_ptr<ICommand> command) {
        _off_command = std::move(command);
    }

    void press_on_button() {
        _on_command->execute();
    }

    void press_off_button() {
        _off_command->execute();
    }

private:
    std::shared_ptr<ICommand> _on_command;
    std::shared_ptr<ICommand> _off_command;
};

class Bulb {
public:
    void light_on() {
        std::cout << "Light On!" << std::endl;
    }

    void light_off() {
        std::cout << "Light Off!" << std::endl;
    }
};

class LightOnCommand : public ICommand {
public:
    explicit LightOnCommand(Bulb &bulb) : _bulb(bulb) {
    }

    void execute() override {
        _bulb.light_on();
    }

private:
    Bulb &_bulb;
};

class LightOffCommand : public ICommand {
public:
    explicit LightOffCommand(Bulb &bulb) : _bulb(bulb) {
    }

    void execute() override {
        _bulb.light_off();
    }

private:
    Bulb &_bulb;
};

int main() {
    Bulb bulb;
    RemoteController controller;
    controller.set_on_command(std::make_shared<LightOnCommand>(bulb));
    controller.set_off_command(std::make_shared<LightOffCommand>(bulb));
    controller.press_on_button();
    controller.press_off_button();
    return 0;
}
